src/vast-client-serializer.js

```diff
@@ -0,0 +1,703 @@
+   * @param {Array[Object]} creativeExtensions An array of creativeExtensions objects
+   * @returns {Object} An array of CreativeExtension nodes
+   */
+  buildCreativeExtensions(creativeExtensions) {
+    if (!creativeExtensions) {
+      return null
+    }
+
+    return {
+      'CreativeExtension': creativeExtensions.map(
+        (creativeExtension) => this.buildExtension(creativeExtension)
+      ),
+    }
+  }
+
+  /**
+   * Build Creative node from it's type
+   * @param {Object} creative A creative object
+   * @returns {Object} Either a Linear, NonLinearAds, CompanionAds node
+   */
+  buildCreativeByType(creative) {
+    switch (creative.type) {
+      case 'linear':
+        return {
+          'Linear': this.buildLinear(creative)
+        }
+      case 'nonlinear':
+        return {
+          'NonLinearAds': this.buildNonLinearAds(creative)
+        }
+      case 'companion':
+        return {
+          '@required': creative.required,
+          'CompanionAds': this.buildCompanionAds(creative)
+        }
+    }
+  }
+
+  /**
+   * Build Linear node from creative
+   * @param {Object} creative A creative object
+   * @returns {Object} A Linear node
+   */
+  buildLinear(creative) {
+    return {
+      '@skipoffset': this.convertToHHMMSS(creative.skipDelay),
+      'Duration': this.convertToHHMMSS(creative.duration),
+      'AdParameters': this.buildAdParameters(creative.adParameters),
+      'MediaFiles': this.buildMediafiles(
+        creative.mediaFiles,
+        creative.mezzanine,
+        creative.interactiveCreativeFile,
+        creative.closedCaptionFiles
+      ),
+      'VideoClicks': this.buildVideoClicks(
+        creative.videoClickThroughURLTemplate,
+        creative.videoClickTrackingURLTemplates,
+        creative.videoCustomClickURLTemplates,
+      ),
+      'TrackingEvents': this.buildTrackingEvents(creative.trackingEvents),
+      'Icons': this.buildIcons(creative.icons),
+    }
+  }
+
+  /**
+   * Build AdParameters node
+   * @param {String} adParameters The adParameters string
+   * @param {Boolean} xmlEncoded The xmlEncoded parameter
+   * @returns {Object} A AdParameters node
+   */
+  buildAdParameters(adParameters, xmlEncoded = null) {
+    if (!adParameters) {
+      return null
+    }
+
+    return {
+      '@xmlEncoded': xmlEncoded,
+      '#': this.wrapCDATA(adParameters),
+    }
+  }
+
+  /**
+   * Build VideoClicks parameter node
+   * @param {Object} clickThrough The clickThrough object
+   * @param {Array[Object]} clickTrackings The array of clickTracking objects
+   * @param {Array[Object]} customClicks The array of customClicks objects
+   * @returns {Object} A VideoClicks node
+   */
+  buildVideoClicks(clickThrough, clickTrackings, customClicks) {
+    let videoClicks = {}
+
+    if (clickThrough) {
+      const clickThroughObj = clickThrough === Object ? clickThrough : {
+        url: clickThrough,
+      }
+
+      videoClicks = {
+        ...videoClicks,
```

Comment:
IMO the spread can be removed as it's defined as empty

Reply:
It's to avoid bugs in the future, if tomorrow we mutate the object before, we won't risk this code to override it

Reply:
I don't have a strong opinion on it, but we could move l.415 to l.422 since we don't use it between thoses line. Like so we make sure that we wont mutate it before 😉 
```
let videoClicks  = {
   'ClickThrough': {
   ...
   ```

Reply:
clickThrough can be null, so it might break the next lines since videoClicks won't be defined

Reply:
My bad, i forgot that we where in the if scope 🤦‍♂️ 
All good then 👍